src/detectors/builtin/shortCircuitCondition.ts

```diff
@@ -0,0 +1,170 @@
+ * placing constant or cheaper conditions first can prevent unnecessary execution
+ * of expensive operations when the result is already determined.
+ *
+ * ## Example
+ * ```tact
+ * // Bad: Expensive operation is always executed
+ * if (expensive_function() && constant_false) {
+ *   // ...
+ * }
+ * ```
+ *
+ * Use instead:
+ * ```tact
+ * // Good: Expensive operation is skipped when constant_false is false
+ * if (constant_false && expensive_function()) {
+ *   // ...
+ * }
+ * ```
+ */
+export class ShortCircuitCondition extends ASTDetector {
+  severity = Severity.LOW;
+
+  async check(cu: CompilationUnit): Promise<MistiTactWarning[]> {
+    return Array.from(cu.ast.getFunctions()).reduce(
+      (acc, fun) => acc.concat(this.checkFunction(fun)),
+      [] as MistiTactWarning[],
+    );
+  }
+
+  private checkFunction(fun: any): MistiTactWarning[] {
+    return foldStatements(
+      fun,
+      (acc, stmt) => {
+        forEachExpression(stmt, (expr) => {
+          if (
+            expr.kind === "op_binary" &&
+            (expr.op === "&&" || expr.op === "||")
+          ) {
+            const leftConst = this.isConstantExpression(expr.left);
+            const rightConst = this.isConstantExpression(expr.right);
+            const leftExpensive = this.containsExpensiveCall(expr.left);
+            const rightExpensive = this.containsExpensiveCall(expr.right);
+            if (
+              expr.op === "&&" &&
+              leftExpensive &&
+              !rightExpensive &&
+              !leftConst &&
+              !rightConst
+            ) {
+              acc.push(
+                this.makeWarning(
+                  `Consider reordering: Move expensive function call to the end.`,
+                  expr.loc,
+                  {
+                    suggestion: `Place cheaper conditions on the left to leverage short-circuiting.`,
+                  },
+                ),
+              );
+            } else if (
+              expr.op === "||" &&
+              leftExpensive &&
+              !rightExpensive &&
+              !leftConst &&
+              !rightConst
+            ) {
+              acc.push(
+                this.makeWarning(
+                  `Consider reordering: Move expensive function call to the end.`,
+                  expr.loc,
+                  {
+                    suggestion: `Place cheaper conditions on the left to leverage short-circuiting.`,
+                  },
+                ),
+              );
+            } else if (
+              (expr.op === "&&" && !leftConst && rightConst) ||
+              (expr.op === "||" && leftConst && !rightConst)
+            ) {
+              acc.push(
+                this.makeWarning(
+                  `Consider reordering: Move constant to the left.`,
+                  expr.loc,
+                  {
+                    suggestion: `Reorder to optimize ${expr.op} condition short-circuiting`,
```

Comment:
Please pretty-print the suggestion. Otherwise, it might not be clear for the user who has never heard about short-circuiting.